stl/inc/algorithm

```diff
@@ -7594,9 +7599,78 @@ namespace ranges {
+            requires sortable<_It, _Pr, _Pj>
+        _NODISCARD constexpr _It operator()(_It _First, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            // clang-format on
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst = _Get_unwrapped(_STD move(_First));
+            auto _ULast  = _Get_final_iterator_unwrapped<_It>(_UFirst, _STD move(_Last));
+            _Seek_wrapped(_First, _ULast);
+            const auto _Count = _ULast - _UFirst;
+            _Sort_common(_STD move(_UFirst), _STD move(_ULast), _Count, _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _First;
+        }
+
+        // clang-format off
+        template <random_access_range _Rng, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<iterator_t<_Rng>, _Pr, _Pj>
+        _NODISCARD constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            // clang-format on
+            auto _ULast       = _Get_final_iterator_unwrapped(_Range);
+            auto _UFirst      = _Ubegin(_Range);
+            const auto _Count = _ULast - _UFirst;
+            _Sort_common(_STD move(_UFirst), _ULast, _Count, _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _Rewrap_iterator(_Range, _STD move(_ULast));
+        }
+
+    private:
+        template <class _It, class _Pr, class _Pj>
+        _NODISCARD static constexpr void _Sort_common(
+            _It _First, _It _Last, iter_difference_t<_It> _Ideal, _Pr _Pred, _Pj _Proj) {
+            // sort [_First, _Last) with respect to _Pred and _Proj
+            _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+
+            for (;;) {
+                if (_Last - _First <= _ISORT_MAX) { // small
+                    _RANGES _Insertion_sort_common(_STD move(_First), _STD move(_Last), _Pred, _Proj);
+                    return;
+                }
+
+                if (_Ideal <= 0) { // heap sort if too many divisions
+                    _RANGES _Make_heap_common(_First, _Last, _Pred, _Proj);
+                    _RANGES _Sort_heap_common(_STD move(_First), _STD move(_Last), _Pred, _Proj);
+                    return;
+                }
+
+                // divide and conquer by quicksort
+                auto [_Mid_first, _Mid_last] =
+                    _RANGES _Partition_by_median_guess_unchecked(_First, _Last, _Pred, _Proj);
+
+                _Ideal = (_Ideal >> 1) + (_Ideal >> 2); // allow 1.5 log2(N) divisions
+
+                if (_Mid_first - _First < _Last - _Mid_last) { // loop on second half
+                    _Sort_common(_First, _Mid_first, _Ideal, _Pred, _Proj);
+                    _First = _Mid_last;
+                } else { // loop on first half
+                    _Sort_common(_Mid_last, _Last, _Ideal, _Pred, _Proj);
+                    _Last = _Mid_first;
+                }

```

Comment:
As `[_Mid_first, _Mid_last]` are about to go out of scope, should their final uses here be moved-from?

Reply:
Yes, that is more consistent with the policies I've been using for discretionary iterator moves.